File: backend/internal/adapters/http/billing_estimate.go
package httpadapter

import (
	"context"
	"errors"
	"net/http"

	appbillingestimate "power-iot-backend/internal/application/billingestimate"
	corebillingenergy "power-iot-backend/internal/core/billingenergy"
	corebillingestimate "power-iot-backend/internal/core/billingestimate"

	"github.com/gin-gonic/gin"
)

type BillingEstimateQuery interface {
	Find(context.Context, uint, uint, string) (appbillingestimate.Estimate, error)
}

func NewBillingEstimateHandler(query BillingEstimateQuery) gin.HandlerFunc {
	return func(c *gin.Context) {
		identity, ok := IdentityFromGin(c)
		if !ok || query == nil {
			WritePublicError(c, ErrUnauthorized)
			return
		}
		userID, err := parseExternalID(identity.UserID)
		if err != nil {
			WritePublicError(c, ErrUnauthorized)
			return
		}
		shopID, err := parseExternalID(c.Param("shopId"))
		if err != nil {
			WritePublicError(c, ErrValidation)
			return
		}
		month := c.Query("month")
		if month == "" {
			WritePublicError(c, ErrValidation)
			return
		}
		result, err := query.Find(c.Request.Context(), userID, shopID, month)
		if err != nil {
			if errors.Is(err, corebillingestimate.ErrEstimateAccess) {
				WritePublicError(c, ErrForbidden)
				return
			}
			if errors.Is(err, corebillingenergy.ErrInvalidBillingMonth) || errors.Is(err, corebillingestimate.ErrInvalidCalculationInput) {
				WritePublicError(c, ErrValidation)
				return
			}
			WritePublicError(c, err)
			return
		}
		c.JSON(http.StatusOK, estimateResponse(result))
	}
}

type estimateResponseBody struct {
	Status   string                   `json:"status"`
	Month    string                   `json:"month"`
	Period   estimatePeriodResponse   `json:"period"`
	Shop     estimateShopResponse     `json:"shop"`
	Tariff   estimateTariffResponse   `json:"tariff"`
	RateSet  estimateRateSetResponse  `json:"rateSet"`
	Energy   estimateEnergyResponse   `json:"energy"`
	Tiers    []estimateTierResponse   `json:"tiers"`
	Charges  *estimateChargesResponse `json:"charges"`
	Warnings []string                 `json:"warnings"`
}

type estimatePeriodResponse struct {
	Start    *string `json:"start,omitempty"`
	End      *string `json:"end,omitempty"`
	Timezone string  `json:"timezone"`
}

type estimateShopResponse struct {
	ID   string `json:"id"`
	Code string `json:"code"`
	Name string `json:"name"`
}

type estimateTariffResponse struct {
	ElectricityTariff string  `json:"electricityTariff"`
	PlanCode          string  `json:"planCode"`
	UsageClass        *string `json:"usageClass,omitempty"`
	Season            string  `json:"season"`
}

type estimateRateSetResponse struct {
	Version     string `json:"version"`
	Currency    string `json:"currency"`
	IncludesTax bool   `json:"includesTax"`
}

type estimateEnergyResponse struct {
	UsageKwh                *string `json:"usageKwh"`
	ExpectedDurationSeconds int64   `json:"expectedDurationSeconds"`
	ObservedDurationSeconds int64   `json:"observedDurationSeconds"`
	Coverage                *string `json:"coverage"`
}

type estimateTierResponse struct {
	FromKwh    string  `json:"fromKwh"`
	ToKwh      *string `json:"toKwh"`
	UsageKwh   string  `json:"usageKwh"`
	RatePerKwh string  `json:"ratePerKwh"`
	Subtotal   string  `json:"subtotal"`
}

type estimateChargesResponse struct {
	EnergyCharge            string `json:"energyCharge"`
	MinimumMonthlyCharge    string `json:"minimumMonthlyCharge"`
	MinimumChargeAdjustment string `json:"minimumChargeAdjustment"`
	EstimatedTotal          string `json:"estimatedTotal"`
}

func estimateResponse(result appbillingestimate.Estimate) estimateResponseBody {
	var start, end *string
	if !result.Period.Start.IsZero() {
		value := result.Period.Start.UTC().Format("2006-01-02T15:04:05Z07:00")
		start = &value
	}
	if !result.Period.End.IsZero() {
		value := result.Period.End.UTC().Format("2006-01-02T15:04:05Z07:00")
		end = &value
	}
	tiers := make([]estimateTierResponse, 0, len(result.Tiers))
	for _, tier := range result.Tiers {
		tiers = append(tiers, estimateTierResponse{FromKwh: tier.FromKwh, ToKwh: tier.ToKwh, UsageKwh: tier.UsageKwh, RatePerKwh: tier.RatePerKwh, Subtotal: tier.Subtotal})
	}
	warnings := make([]string, 0, len(result.Warnings))
	for _, warning := range result.Warnings {
		warnings = append(warnings, string(warning))
	}
	var charges *estimateChargesResponse
	if result.Charges != nil {
		charges = &estimateChargesResponse{EnergyCharge: result.Charges.EnergyCharge, MinimumMonthlyCharge: result.Charges.MinimumMonthlyCharge, MinimumChargeAdjustment: result.Charges.MinimumChargeAdjustment, EstimatedTotal: result.Charges.EstimatedTotal}
	}
	return estimateResponseBody{
		Status: string(result.Status), Month: result.Month,
		Period:  estimatePeriodResponse{Start: start, End: end, Timezone: result.Period.Timezone},
		Shop:    estimateShopResponse{ID: result.Shop.ID, Code: result.Shop.Code, Name: result.Shop.Name},
		Tariff:  estimateTariffResponse{ElectricityTariff: result.Tariff.ElectricityTariff, PlanCode: result.Tariff.PlanCode, UsageClass: result.Tariff.UsageClass, Season: result.Tariff.Season},
		RateSet: estimateRateSetResponse{Version: result.RateSet.Version, Currency: result.RateSet.Currency, IncludesTax: result.RateSet.IncludesTax},
		Energy:  estimateEnergyResponse{UsageKwh: result.Energy.UsageKwh, ExpectedDurationSeconds: result.Energy.ExpectedDurationSeconds, ObservedDurationSeconds: result.Energy.ObservedDurationSeconds, Coverage: result.Energy.Coverage},
		Tiers:   tiers, Charges: charges, Warnings: warnings,
	}
}

func RegisterBillingEstimateRoute(router gin.IRouter, authenticator AccessTokenAuthenticator, query BillingEstimateQuery) {
	if router == nil {
		return
	}
	router.GET("/api/v1/shops/:shopId/billing/estimate", AuthenticationMiddleware(authenticator), NewBillingEstimateHandler(query))
}

File: mobile/lib/features/billing/domain/models/billing_estimate.dart
enum BillingEstimateStatus {
  completeEstimate('COMPLETE_ESTIMATE'),
  partialDataEstimate('PARTIAL_DATA_ESTIMATE'),
  noData('NO_DATA'),
  configurationRequired('CONFIGURATION_REQUIRED'),
  unsupportedTariff('UNSUPPORTED_TARIFF'),
  unsupportedPeriod('UNSUPPORTED_PERIOD'),
  rateNotFound('RATE_NOT_FOUND');

  const BillingEstimateStatus(this.code);
  final String code;

  static BillingEstimateStatus fromCode(Object? value) {
    if (value is! String) {
      throw const FormatException('Invalid estimate status');
    }
    return BillingEstimateStatus.values.firstWhere(
      (status) => status.code == value,
      orElse: () => throw const FormatException('Invalid estimate status'),
    );
  }
}

class BillingEstimate {
  const BillingEstimate({
    required this.status,
    required this.month,
    required this.period,
    required this.shop,
    required this.tariff,
    required this.rateSet,
    required this.energy,
    required this.tiers,
    required this.charges,
    required this.warnings,
  });

  final BillingEstimateStatus status;
  final String month;
  final BillingEstimatePeriod period;
  final BillingEstimateShop shop;
  final BillingEstimateTariff tariff;
  final BillingEstimateRateSet rateSet;
  final BillingEstimateEnergy energy;
  final List<BillingEstimateTier> tiers;
  final BillingEstimateCharges? charges;
  final List<String> warnings;

  bool get hasAmount =>
      charges != null &&
      (status == BillingEstimateStatus.completeEstimate ||
          status == BillingEstimateStatus.partialDataEstimate);

  factory BillingEstimate.fromJson(Object? value) {
    if (value is! Map) throw const FormatException('Invalid billing estimate');
    final map = value.map((key, value) => MapEntry(key.toString(), value));
    const keys = {
      'status',
      'month',
      'period',
      'shop',
      'tariff',
      'rateSet',
      'energy',
      'tiers',
      'charges',
      'warnings',
    };
    if (map.length != keys.length ||
        !map.keys.every(keys.contains) ||
        map['month'] is! String ||
        map['tiers'] is! List ||
        map['warnings'] is! List) {
      throw const FormatException('Invalid billing estimate');
    }
    return BillingEstimate(
      status: BillingEstimateStatus.fromCode(map['status']),
      month: map['month'] as String,
      period: BillingEstimatePeriod.fromJson(map['period']),
      shop: BillingEstimateShop.fromJson(map['shop']),
      tariff: BillingEstimateTariff.fromJson(map['tariff']),
      rateSet: BillingEstimateRateSet.fromJson(map['rateSet']),
      energy: BillingEstimateEnergy.fromJson(map['energy']),
      tiers: List.unmodifiable(
        (map['tiers'] as List).map(BillingEstimateTier.fromJson),
      ),
      charges: map['charges'] == null
          ? null
          : BillingEstimateCharges.fromJson(map['charges']),
      warnings: List.unmodifiable(
        (map['warnings'] as List).map((value) {
          if (value is! String) throw const FormatException('Invalid warning');
          return value;
        }),
      ),
    );
  }
}

Map<String, Object?> _map(Object? value, String name) {
  if (value is! Map) throw FormatException('Invalid $name');
  return value.map((key, value) => MapEntry(key.toString(), value));
}

class BillingEstimatePeriod {
  const BillingEstimatePeriod({
    required this.start,
    required this.end,
    required this.timezone,
  });
  final String? start;
  final String? end;
  final String timezone;

  factory BillingEstimatePeriod.fromJson(Object? value) {
    final map = _map(value, 'period');
    if (map['start'] != null && map['start'] is! String ||
        map['end'] != null && map['end'] is! String ||
        map['timezone'] is! String) {
      throw const FormatException('Invalid estimate period');
    }
    return BillingEstimatePeriod(
      start: map['start'] as String?,
      end: map['end'] as String?,
      timezone: map['timezone'] as String,
    );
  }
}

class BillingEstimateShop {
  const BillingEstimateShop({
    required this.id,
    required this.code,
    required this.name,
  });
  final String id;
  final String code;
  final String name;

  factory BillingEstimateShop.fromJson(Object? value) {
    final map = _map(value, 'shop');
    if (map['id'] is! String ||
        map['code'] is! String ||
        map['name'] is! String) {
      throw const FormatException('Invalid estimate shop');
    }
    return BillingEstimateShop(
      id: map['id'] as String,
      code: map['code'] as String,
      name: map['name'] as String,
    );
  }
}

class BillingEstimateTariff {
  const BillingEstimateTariff({
    required this.electricityTariff,
    required this.planCode,
    required this.usageClass,
    required this.season,
  });
  final String electricityTariff;
  final String planCode;
  final String? usageClass;
  final String season;

  factory BillingEstimateTariff.fromJson(Object? value) {
    final map = _map(value, 'tariff');
    if (map['electricityTariff'] is! String ||
        map['planCode'] is! String ||
        map['season'] is! String ||
        map['usageClass'] != null && map['usageClass'] is! String) {
      throw const FormatException('Invalid estimate tariff');
    }
    return BillingEstimateTariff(
      electricityTariff: map['electricityTariff'] as String,
      planCode: map['planCode'] as String,
      usageClass: map['usageClass'] as String?,
      season: map['season'] as String,
    );
  }
}

class BillingEstimateRateSet {
  const BillingEstimateRateSet({
    required this.version,
    required this.currency,
    required this.includesTax,
  });
  final String version;
  final String currency;
  final bool includesTax;

  factory BillingEstimateRateSet.fromJson(Object? value) {
    final map = _map(value, 'rate set');
    if (map['version'] is! String ||
        map['currency'] is! String ||
        map['includesTax'] is! bool) {
      throw const FormatException('Invalid estimate rate set');
    }
    return BillingEstimateRateSet(
      version: map['version'] as String,
      currency: map['currency'] as String,
      includesTax: map['includesTax'] as bool,
    );
  }
}

class BillingEstimateEnergy {
  const BillingEstimateEnergy({
    required this.usageKwh,
    required this.expectedDurationSeconds,
    required this.observedDurationSeconds,
    required this.coverage,
  });
  final String? usageKwh;
  final int expectedDurationSeconds;
  final int observedDurationSeconds;
  final String? coverage;

  factory BillingEstimateEnergy.fromJson(Object? value) {
    final map = _map(value, 'energy');
    if (map['usageKwh'] != null && map['usageKwh'] is! String ||
        map['coverage'] != null && map['coverage'] is! String ||
        map['expectedDurationSeconds'] is! int ||
        map['observedDurationSeconds'] is! int) {
      throw const FormatException('Invalid estimate energy');
    }
    return BillingEstimateEnergy(
      usageKwh: map['usageKwh'] as String?,
      expectedDurationSeconds: map['expectedDurationSeconds'] as int,
      observedDurationSeconds: map['observedDurationSeconds'] as int,
      coverage: map['coverage'] as String?,
    );
  }
}

class BillingEstimateTier {
  const BillingEstimateTier({
    required this.fromKwh,
    required this.toKwh,
    required this.usageKwh,
    required this.ratePerKwh,
    required this.subtotal,
  });
  final String fromKwh;
  final String? toKwh;
  final String usageKwh;
  final String ratePerKwh;
  final String subtotal;

  factory BillingEstimateTier.fromJson(Object? value) {
    final map = _map(value, 'tier');
    if (map['fromKwh'] is! String ||
        map['usageKwh'] is! String ||
        map['ratePerKwh'] is! String ||
        map['subtotal'] is! String ||
        map['toKwh'] != null && map['toKwh'] is! String) {
      throw const FormatException('Invalid estimate tier');
    }
    return BillingEstimateTier(
      fromKwh: map['fromKwh'] as String,
      toKwh: map['toKwh'] as String?,
      usageKwh: map['usageKwh'] as String,
      ratePerKwh: map['ratePerKwh'] as String,
      subtotal: map['subtotal'] as String,
    );
  }
}

class BillingEstimateCharges {
  const BillingEstimateCharges({
    required this.energyCharge,
    required this.minimumMonthlyCharge,
    required this.minimumChargeAdjustment,
    required this.estimatedTotal,
  });
  final String energyCharge;
  final String minimumMonthlyCharge;
  final String minimumChargeAdjustment;
  final String estimatedTotal;

  factory BillingEstimateCharges.fromJson(Object? value) {
    final map = _map(value, 'charges');
    if (map['energyCharge'] is! String ||
        map['minimumMonthlyCharge'] is! String ||
        map['minimumChargeAdjustment'] is! String ||
        map['estimatedTotal'] is! String) {
      throw const FormatException('Invalid estimate charges');
    }
    return BillingEstimateCharges(
      energyCharge: map['energyCharge'] as String,
      minimumMonthlyCharge: map['minimumMonthlyCharge'] as String,
      minimumChargeAdjustment: map['minimumChargeAdjustment'] as String,
      estimatedTotal: map['estimatedTotal'] as String,
    );
  }
}
